Yoast/wordpress-seo
# Add recommended replacement variables

## Summary

This PR can be summarized in the following changelog entry:

* Adds recommended snippet variables for templates depending on the context. The `wpseo_recommended_replace_vars` filter is added. Which gives the possibility to add or alter the recommended snippet variables.

## Props

Thanks @andizer !

## Relevant technical choices:

* Adds a class to determine the page type and which recommended replace variables go with that type.
* Determines the recommended_replace_vars in the replace_vars localization.
* In the scrapers these get added to the args.
* In edit.js they are added to the snippet editor props.
* Adds a filter named `wpseo_recommended_replace_vars` for plugins to add / alter recommendations.
* Recommended replace vars for a custom post type: `sitename`, `title`, `sep`.
* Recommended replace vars for a custom post type archive: `sitename`, `title`, `sep` and `page`.

## Answered Questions

* Do they have to be in the same order as in the issue? **Answer: no**.
* The Woo shipping classes are not on their own page with a Yoast metabox. And it does not look like they have a XML sitemap either. Does this need to be in this list? **Answer: no**.
* If a page is also the homepage, do we want the recommended replacement variables of the homepage instead of the page ones? Presumed yes. **Answer: yes**.

## Test instructions

This PR can be tested by following these steps:

* As specified in the issue, there should be different recommended replacement variables per context. See the issue for the list.
* Check if they equal the items that show as suggestions when you just insert a `%`.
<img width="639" alt="sitename title primary_category sep" src="https://user-images.githubusercontent.com/35524806/41402364-7094fef8-6fc2-11e8-95b6-afc196577a9b.png">
* Check that it still filters on all the replacement variables when typing.
<img width="638" alt="normal filter" src="https://user-images.githubusercontent.com/35524806/41402509-c28250d0-6fc2-11e8-87ff-9ff4f49bcfc3.png">
* For in the Settings -> Appearance it can be handy to show the actual used page type too. This can be done in the `SettingsReplacementVariableEditors.js` file on lines 45 and 75, by outputting the `reactReplacevarPageType` variable.

## Quality assurance


Fixes #9888
